File: cpp-solutions/227.cpp
class Solution {
public:
    int calculate(string s) {
        int res=0, tmp = 0, num;
        char ope = '+';
        stack<int> sta;
        for (int i=0; i<s.size(); i++){
            if (isdigit(s[i])){
                tmp = tmp*10 + s[i] - '0';
            }
            if (!isdigit(s[i]) && !isspace(s[i]) || i==s.size()-1){
                switch (ope){
                    case '+':
                        sta.push(tmp);
                        break;
                    case '-':
                        sta.push(-tmp);
                        break;
                    case '*':
                        num = sta.top() * tmp;
                        sta.pop();
                        sta.push(num);
                        break;
                    case '/':
                        num = sta.top() / tmp;
                        sta.pop();
                        sta.push(num);
                        break;
                }
                ope = s[i];
                tmp = 0;
            }
        }
        
        while(!sta.empty()){
            res += sta.top();
            sta.pop();
        }
        return res;
    }
};

File: golang/227/227.go
package main

func calculate(s string) int {
	stack := make([]int, 0)
	num := 0
	sign := '+'

	for i, v := range s {
		if isDigit(v) {
			num = num*10 + int(v-'0')
		}

		if (!isDigit(v) && v != ' ') || i == len(s)-1 {
			switch sign {
			case '+':
				stack = append(stack, num)

			case '-':
				stack = append(stack, -1*num)

			case '*':
				pre := stack[len(stack)-1]
				stack = stack[:len(stack)-1]
				stack = append(stack, pre*num)

			case '/':
				pre := stack[len(stack)-1]
				stack = stack[:len(stack)-1]
				stack = append(stack, pre/num)
			}

			sign = v
			num = 0
		}
	}

	res := 0
	for len(stack) > 0 {
		res = res + stack[len(stack)-1]
		stack = stack[:len(stack)-1]
	}

	return res
}

func isDigit(s rune) bool {
	return s >= '0' && s <= '9'
}
